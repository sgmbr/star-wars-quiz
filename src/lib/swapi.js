const BASE_URL = 'https://swapi.co/api/'

function checkStatus (response) {
  if (response.status >= 200 && response.status < 300) {
    return Promise.resolve(response)
  } else {
    return Promise.reject(new Error(response.statusText))
  }
}

function json (response) {
  return response.json()
}

function fetchSwapi(url) {
  return (
    fetch(url)
      .then(checkStatus)
      .then(json)
      .catch(err => {
        console.log('Request failed', err)
      })
  )
}

async function getAllEntries (url, entries) {
  entries = entries || []
  const data = await fetchSwapi(url)
  entries = entries.concat(data.results)

  if (data.next) {
    return getAllEntries(data.next, entries)
  } else {
    return entries
  }
}

function getAllEntriesUrls (url) {
  return getAllEntries(url)
          .then(entries => entries.map(entry => entry.url))
}

export { BASE_URL, fetchSwapi, getAllEntriesUrls }
