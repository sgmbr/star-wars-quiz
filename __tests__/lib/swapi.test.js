import { BASE_URL, fetchSwapi, getAllEntriesUrls } from '../../src/lib/swapi.js'
import 'whatwg-fetch'

describe('fetchSwapi', () => {
  test('should return Luke Skywalker', () => {
    expect.assertions(1)
    return expect(fetchSwapi(`${BASE_URL}people/1/`)).resolves.toMatchSnapshot()
  })
})

describe('getAllEntriesUrls', () => {
  let originalTimeout
  beforeEach(() => {
    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000
  })

  it('should return all the urls for starships', () => {
    expect.assertions(1)
    return expect(getAllEntriesUrls(`${BASE_URL}starships`)).resolves.toMatchSnapshot()
  })

  it('should return all the urls for people', () => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000
    expect.assertions(1)
    return expect(getAllEntriesUrls(`${BASE_URL}people`)).resolves.toMatchSnapshot()
  })

  afterEach(() => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout
  })
})
