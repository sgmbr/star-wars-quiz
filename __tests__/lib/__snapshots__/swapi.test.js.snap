// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchSwapi should return Luke Skywalker 1`] = `
Object {
  "birth_year": "19BBY",
  "created": "2014-12-09T13:50:51.644000Z",
  "edited": "2014-12-20T21:17:56.891000Z",
  "eye_color": "blue",
  "films": Array [
    "https://swapi.co/api/films/2/",
    "https://swapi.co/api/films/6/",
    "https://swapi.co/api/films/3/",
    "https://swapi.co/api/films/1/",
    "https://swapi.co/api/films/7/",
  ],
  "gender": "male",
  "hair_color": "blond",
  "height": "172",
  "homeworld": "https://swapi.co/api/planets/1/",
  "mass": "77",
  "name": "Luke Skywalker",
  "skin_color": "fair",
  "species": Array [
    "https://swapi.co/api/species/1/",
  ],
  "starships": Array [
    "https://swapi.co/api/starships/12/",
    "https://swapi.co/api/starships/22/",
  ],
  "url": "https://swapi.co/api/people/1/",
  "vehicles": Array [
    "https://swapi.co/api/vehicles/14/",
    "https://swapi.co/api/vehicles/30/",
  ],
}
`;

exports[`getAllEntriesUrls should return all the urls for people 1`] = `
Array [
  "https://swapi.co/api/people/1/",
  "https://swapi.co/api/people/2/",
  "https://swapi.co/api/people/3/",
  "https://swapi.co/api/people/4/",
  "https://swapi.co/api/people/5/",
  "https://swapi.co/api/people/6/",
  "https://swapi.co/api/people/7/",
  "https://swapi.co/api/people/8/",
  "https://swapi.co/api/people/9/",
  "https://swapi.co/api/people/10/",
  "https://swapi.co/api/people/11/",
  "https://swapi.co/api/people/12/",
  "https://swapi.co/api/people/13/",
  "https://swapi.co/api/people/14/",
  "https://swapi.co/api/people/15/",
  "https://swapi.co/api/people/16/",
  "https://swapi.co/api/people/18/",
  "https://swapi.co/api/people/19/",
  "https://swapi.co/api/people/20/",
  "https://swapi.co/api/people/21/",
  "https://swapi.co/api/people/22/",
  "https://swapi.co/api/people/23/",
  "https://swapi.co/api/people/24/",
  "https://swapi.co/api/people/25/",
  "https://swapi.co/api/people/26/",
  "https://swapi.co/api/people/27/",
  "https://swapi.co/api/people/28/",
  "https://swapi.co/api/people/29/",
  "https://swapi.co/api/people/30/",
  "https://swapi.co/api/people/31/",
  "https://swapi.co/api/people/32/",
  "https://swapi.co/api/people/33/",
  "https://swapi.co/api/people/34/",
  "https://swapi.co/api/people/36/",
  "https://swapi.co/api/people/37/",
  "https://swapi.co/api/people/38/",
  "https://swapi.co/api/people/39/",
  "https://swapi.co/api/people/40/",
  "https://swapi.co/api/people/41/",
  "https://swapi.co/api/people/42/",
  "https://swapi.co/api/people/43/",
  "https://swapi.co/api/people/44/",
  "https://swapi.co/api/people/45/",
  "https://swapi.co/api/people/46/",
  "https://swapi.co/api/people/48/",
  "https://swapi.co/api/people/49/",
  "https://swapi.co/api/people/50/",
  "https://swapi.co/api/people/51/",
  "https://swapi.co/api/people/52/",
  "https://swapi.co/api/people/53/",
  "https://swapi.co/api/people/54/",
  "https://swapi.co/api/people/55/",
  "https://swapi.co/api/people/56/",
  "https://swapi.co/api/people/57/",
  "https://swapi.co/api/people/58/",
  "https://swapi.co/api/people/59/",
  "https://swapi.co/api/people/60/",
  "https://swapi.co/api/people/61/",
  "https://swapi.co/api/people/62/",
  "https://swapi.co/api/people/63/",
  "https://swapi.co/api/people/64/",
  "https://swapi.co/api/people/65/",
  "https://swapi.co/api/people/66/",
  "https://swapi.co/api/people/67/",
  "https://swapi.co/api/people/68/",
  "https://swapi.co/api/people/69/",
  "https://swapi.co/api/people/70/",
  "https://swapi.co/api/people/71/",
  "https://swapi.co/api/people/72/",
  "https://swapi.co/api/people/73/",
  "https://swapi.co/api/people/74/",
  "https://swapi.co/api/people/47/",
  "https://swapi.co/api/people/75/",
  "https://swapi.co/api/people/76/",
  "https://swapi.co/api/people/77/",
  "https://swapi.co/api/people/78/",
  "https://swapi.co/api/people/79/",
  "https://swapi.co/api/people/80/",
  "https://swapi.co/api/people/81/",
  "https://swapi.co/api/people/82/",
  "https://swapi.co/api/people/83/",
  "https://swapi.co/api/people/84/",
  "https://swapi.co/api/people/85/",
  "https://swapi.co/api/people/86/",
  "https://swapi.co/api/people/87/",
  "https://swapi.co/api/people/88/",
  "https://swapi.co/api/people/35/",
]
`;

exports[`getAllEntriesUrls should return all the urls for starships 1`] = `
Array [
  "https://swapi.co/api/starships/15/",
  "https://swapi.co/api/starships/5/",
  "https://swapi.co/api/starships/9/",
  "https://swapi.co/api/starships/10/",
  "https://swapi.co/api/starships/11/",
  "https://swapi.co/api/starships/12/",
  "https://swapi.co/api/starships/13/",
  "https://swapi.co/api/starships/21/",
  "https://swapi.co/api/starships/22/",
  "https://swapi.co/api/starships/23/",
  "https://swapi.co/api/starships/27/",
  "https://swapi.co/api/starships/28/",
  "https://swapi.co/api/starships/29/",
  "https://swapi.co/api/starships/31/",
  "https://swapi.co/api/starships/39/",
  "https://swapi.co/api/starships/40/",
  "https://swapi.co/api/starships/41/",
  "https://swapi.co/api/starships/43/",
  "https://swapi.co/api/starships/47/",
  "https://swapi.co/api/starships/48/",
  "https://swapi.co/api/starships/49/",
  "https://swapi.co/api/starships/3/",
  "https://swapi.co/api/starships/59/",
  "https://swapi.co/api/starships/61/",
  "https://swapi.co/api/starships/77/",
  "https://swapi.co/api/starships/17/",
  "https://swapi.co/api/starships/32/",
  "https://swapi.co/api/starships/52/",
  "https://swapi.co/api/starships/58/",
  "https://swapi.co/api/starships/63/",
  "https://swapi.co/api/starships/64/",
  "https://swapi.co/api/starships/65/",
  "https://swapi.co/api/starships/66/",
  "https://swapi.co/api/starships/74/",
  "https://swapi.co/api/starships/75/",
  "https://swapi.co/api/starships/2/",
  "https://swapi.co/api/starships/68/",
]
`;
